### PR TITLE
Removes the initial delay seconds and drops the period to 1 second

### LINKED ISSUE #21111: Integrate Startup probe to the deployment
### Description

When the Keycloak application is started inside the StatefulSet, probes for checking the state of the initialization are delayed as we use the liveness and readiness probe with the initial delay set to 20s. It causes major delays. 

We could also leverage the Quarkus Startup probe and integrate it with the Keycloak StatefulSet.

### PR DESCRIPTION
**Pull request by shawkins**
Separated these changes from https://github.com/keycloak/keycloak/pull/21658#discussion_r1270554809

With the values and probe implementations we currently have it doesn't seem to make sense to have a initial delay.  Also setting the period delay to 1 second, while having the total failure time, makes the probe more responsive to passing.

However it seems like there may be some cases where it's taking pods to become ready - such as starting against a database in a different AZ.  

We're currently just relying on the default smallrye liveness and startup probes.  I am not familar enough with how things are functioning to understand if the default startup probe would catch this situation - it seems like it would be tied to the notion of quarkus startup.

What @mabartos was suggesting on #21111 would be to add:

```
        if (!containerBuilder.hasStartupProbe()) {
            containerBuilder.withNewStartupProbe()
                .withPeriodSeconds(1)
                .withFailureThreshold(?)
                .withNewHttpGet()
                .withScheme(protocol)
                .withNewPort(kcPort)
                .withPath(kcRelativePath + "health/started")
                .endHttpGet()
                .endStartupProbe();
        }
```

I'm tested that and it seems to have a neglible impact on performance, so if it is effective in detecting "long startup" conditions, then we should consider adding it and reducing the other failure thresholds.

cc @vmuzikar 

Closes #21111

### CODE REVIEW

**Review by vmuzikar — COMMENTED** [review]
@shawkins Thanks for the PR.

What is the benefit of startup probe for us here? Do we need to delay calling the standard liveness endpoint? AFAIK we don't.

> However it seems like there may be some cases where it's taking pods to become ready - such as starting against a database in a different AZ.
> 
> We're currently just relying on the default smallrye liveness and startup probes. I am not familar enough with how things are functioning to understand if the default startup probe would catch this situation - it seems like it would be tied to the notion of quarkus startup.

We're also adding a [DB check](https://github.com/keycloak/keycloak/blob/bd0f87fc4d281b869ef067afc6bfd8f5eaf3848b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/services/health/KeycloakReadyHealthCheck.java) to the health endpoint.

I'd like to also ask someone from @keycloak/performance for a review.

**Comment by mabartos** [comment]
When we're talking about removing the initial delay and decreasing the check period to 1s, I'm leaning towards not providing the startup probe as it only brings us a little more benefits. 

IMHO, the startup probe would be useful in situations when we might not want to check the readiness/liveness every 1s, but with some bigger period in order to improve performance a little bit. However, as these checks are only some kind of `ping`s, there are no such improvements. Moreover, when we have such a big threshold for the other probes, I don't see much benefits either. We need to pass the readiness check for considering Keycloak to be fully initialized with proper connection to the DB anyway. 

Thus, I'd suggest removing the startup probe, initial delays, and only changing the period of these checks.
@shawkins WDYT?

**Comment by shawkins** [comment]
> What is the benefit of startup probe for us here? Do we need to delay calling the standard liveness endpoint? AFAIK we don't.

This should also answer the comment from @mabartos 

I primarily wanted to ensure that there was not an unintentional regression here.  The prior liveness probe would allow for up to 320 seconds before failing, which is what the added startup probe allows for.  I tried a couple of ways of influencing startup behavior to confirm this - chaos-mesh to influence db latency on the realm import tests (didn't ultimately work out very well), and setting the cpu limit down to 300m.  With the reduced cpu limit keycloak would fail to start without this startup probe.

Also I didn't see any appriciable difference in performance, and finally it better positions us to further refine the failure thresholds - I think we all agree that the current ones are too large.

> We're also adding a [DB check](https://github.com/keycloak/keycloak/blob/bd0f87fc4d281b869ef067afc6bfd8f5eaf3848b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/services/health/KeycloakReadyHealthCheck.java) to the health endpoint.

Yes I saw that one, but didn't see overrides for liveness or startup.

**Comment by ahus1** [comment]
@shawkins / @vmuzikar - here my 2 cents on probes from the storage team perspective: It makes sense to have a startup probe next to the liveness and readiness probes, as they have different purposes and therefore different timings for period and threshold.

* A startup probe provides the container with enough time to do any initialization. For the storage, this include rebalancing Infinispan caches, but also database schema migrations. With a large database, such migrations can take several minutes to complete. While it is good to have a reasonable default here, users might still need to override them when they have large databases. As Keycloak with a pre-built image starts up quite fast, I don't think an initial delay makes sense for Keycloak.
**I recommend a short period (1 second) and a long threshold (600)** so will act after 600 * 1s = 10 minutes

* A liveness probe will kill a container, for example when thread pools are exhausted, or the JVM runs out of memory and GC takes too long, or some connection pool or similar runs out of connections. This wouldn't change from second to second, and shouldn't be too frequent to keep the load on the Kubelet low. When it fails, it shouldn't take too long to kill the pod. As we have a startup probe, there should be no initial delay on the liveness probe.
**I recommend a medium period (10-30 seconds) and a low threshold (3)**  so it will act after 10 * 3 = 30 seconds or 30 * 3 = 90 seconds

* Once there is a start up probe, the only task of the readiness probe is to remove the pod from the loadbalancing if it is not ready for some reason. One could argue to have a lower threshold than the liveness probe, but that would be fine-tuning. Taking a Keycloak out of the loadbalancer usually leads to more load on the other Keycloaks, which might send the cluster spiraling into doom, so let's keep it simple. 
**I recommend the same settings as for the liveness probe**

On the DB check as part of the "health": I'd like to keep it as is.

* The startup of a container will complete only if the DB is available, therefore it doesn't matter much what the startup endpoint of Quarkus tests.  
* The liveness probe should never check the database connection, as database being down should not restart a pod
* The readiness probe can check the connection to the database and then take the Keycloak instance out of rotation until the database becomes available again. It doesn't do much harm except that it prevents a nice error message shown to the user. I can prevent a database to be overloaded when it has problems though.

**Review by ahus1 — CHANGES_REQUESTED** [review]
With my previous comment on probes above, I recommend longer periods for readiness and liveness probes, and lower thresholds as we'll now always have a startup probe present. 

The startup probe threshold could be longer (like 600), but I wouldn't object too much to the current number of 320.

**Comment by shawkins** [comment]
@ahus1 thank you for the feedback

> With my previous comment on probes above, I recommend longer periods for readiness and liveness probes, and lower thresholds as we'll now always have a startup probe present.

This definitely makes more sense.  I tried locally, and we'll have confirmation soon from the pushed changes, that this shouldn't impact test run times - we expect the liveness and readiness checks to succeed immediately after the /started startup probe passes.  

On the whole based upon what we were seeing with #21658 we'll still end up shaving time off of the test runs compared to before due to elimating the initial delay.

> The startup probe threshold could be longer (like 600), but I wouldn't object too much to the current number of 320.

That was only for backwards compatibility with the old probe values, so it seems fine to move forward with 600.

**Review by ahus1 — APPROVED** [review]
LGTM, thank you for adding the startup probe!

**Comment by ahus1** [comment]
@shawkins - there are still two outstanding reviews, so I don't merge it yet. Let me know if you think otherwise.

**Review by mabartos — APPROVED** [review]
@shawkins @ahus1 Thanks guys, LGTM! Seems much more reasonable than before for the prod environment. Agree to have 10min allocated for the startup and the DB migrations. 

These 30s for readiness/liveness checks seems also reasonable to me as there should not be any expected high time-consuming operations after the startup.

I'm totally ok to set these probes like this.

Nitpick: However, we might not be so much informed about our customers' deployments as they might have various needs. The behavior changed quite a lot, and customers might want to tune these probes. We could probably document the workaround for configuring these probes for containers. 

@shawkins @ahus1 WDYT? Might be potentially done in a follow-up.

**Review by vmuzikar — APPROVED** [review]
@ahus1 Thanks for the feedback, what you say makes sense to me.

> Nitpick: However, we might not be so much informed about our customers' deployments as they might have various needs. The behavior changed quite a lot, and customers might want to tune these probes. We could probably document the workaround for configuring these probes for containers.

@mabartos The only way to tweak to probes is currently through the pod templates field which is considered unsupported. Therefore if we officially support tweaking the probes (which I think we should at some point), it should be rather done through a first-class field in the CR. But that's definitely a follow-up WDYT?

**Comment by ahus1** [comment]
@vmuzikar / @mabartos - +1 for tweaking the probes. 

Even more important would be to adjust Java memory limits and pots limits/requests on CPU and memory, which is also only available via the unsupported pod template. 

With the current default, Keycloak doesn't define any CPU or memory requests, which leads to a scheduling on Kubernetes where Keycloak could end up on a node that is already short on CPU and memory and where Keycloak eats up all remaining memory. It might be killed by Linux out-of-memory if the node then comes under memory pressure, or is among the first pods to be evicted by the Kubelet if memory runs low.

So TL;DR: limits and requests on CPU and memory + JVM settings should be even higher on the list.

**Comment by vmuzikar** [comment]
For requests/limits, we already have https://github.com/keycloak/keycloak/issues/20107

For JVM settings, I created https://github.com/keycloak/keycloak/issues/21996

For probes, created https://github.com/keycloak/keycloak/issues/21995

**Comment by shawkins** [comment]
@vmuzikar thanks for adding / linking to the applicable follow-ups.
